OC-4907 Clients and Users should accept public_key

hosh opened the pull request:
- Fixed chef_client code to allow public_key to update from json payload

See:
- https://github.com/opscode/chef_wm/pull/39
- https://github.com/opscode/chef_objects/pull/26
- https://github.com/opscode/chef-pedant/pull/8

hosh committed:
chef_client should accept public_key from json payload

 - Fixed osc_set_values_from_current_client() to only use persisted key if
   it is not present in the payload.
 - Fixed oc_set_values_from_current_client() to only use persisted key if
   it is not present in the payload.

seth commented on src/chef_client.erl:
```diff
@@ -200,14 +200,24 @@ osc_set_values_from_current_client(Client, #chef_client{admin = IsAdmin,
-    set_public_key(C, PublicKey).
+    case chef_object:cert_or_key(C) of
+        {undefined, _} ->
+            set_public_key(C, PublicKey);
```
Or, on second thought, how about adding a fun-head to set_public_key that does nothing if `Public_key =:= undefined`?

hosh committed:
Factored out chef_object:set_key_pair()

 - Consolidated chef_user:set_key_pair() and chef_client:set_key_pair()
   into chef_object:set_key_pair()
 - Changed behavior of chef_object:set_key_pair() so that it only sets
   the private_key field if the PrivateKey parameter is defined (so that
   it will not be emitted back to the API client)
 - Consolidated chef_user:set_public_key() and chef_client:set_public_key()
   into chef_object:set_public_key()
 - Consolidated chef_user:key_version() and chef_client:key_version()
   into chef_object:key_version()
 - Modified chef_client to use the chef_object:set_public_key()
 - Added specs
 - Modified eunit

seth commented on src/chef_object.erl:
```diff
@@ -663,3 +666,49 @@ allowed_keys(ValidKeys, [{Item, _}|Rest]) ->
+            ej:set({<<"certificate">>}, UserEjson, PublicKey)
+    end.
+
+%% Shameless ripped key_version code from chef_client -
```
Time to remove this comment?

seth commented:
Aside from the comment which can now be removed, this looks like a good improvement. MWR.
